[TypeScript][Angular] fix date path parameters (#7476, #7302)

macjohnny opened the pull request:
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/swagger-api/swagger-codegen/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the correct branch: `3.0.0` branch for changes related to OpenAPI spec 3.0. Default: `master`.
- [x] Copied the [technical committee](https://github.com/swagger-api/swagger-codegen/#swagger-codegen-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

fixes #7476, fixes #7302

```yaml
/pet/{birthday}/foo/{id:.+}:
    ...
      parameters:
        - name: birthday
          in: path
          type: string
          format: date-time
        - name: id
          in: path
          type: string
```
now results in 
```typescript
httpClient.get<Array<Pet>>(`${this.basePath}/pet/${encodeURIComponent(String(birthday.toISOString()))}/foo/${encodeURIComponent(String(id_))}`,
```
instead of 
```typescript
httpClient.get<Array<Pet>>(`${this.basePath}/pet/${encodeURIComponent(String(birthday))}/foo/${encodeURIComponent(String(id_))}`,
```

macjohnny commented:
cc @TiFu @taxpon @sebastianhaas @kenisteward @Vrolijkx

macjohnny commented:
successfully tested with the following dummy api endpoint:
```
/pet/{birthday}/foo/{id:.+}:
    get:
      tags:
        - pet
      operationId: findPetsByStatusfoo
      produces:
        - application/xml
        - application/json
      parameters:
        - name: birthday
          in: path
          required: true
          type: string
          format: date-time
        - name: id
          in: path
          required: true
          type: string
      responses:
        '200':
          schema:
            type: array
            items:
              $ref: '#/definitions/Pet'
```

@arjuin can you confirm this resolves issue https://github.com/swagger-api/swagger-codegen/issues/7476 ?

@inthegarage can you confirm this resolves issue https://github.com/swagger-api/swagger-codegen/issues/7302 ?

macjohnny commented:
cc @bedag-moo @defmonk0 @JohannesHoppe

arjuin commented:
It resolves the issue.
Thanks.

defmonk0 commented:
I ran through the code, and it looks fine to me. Was a good catch and simple fix that didn't require heavy reworking.

My only concern would be that this explicitly only works with alphanumeric characters and underscores in parameter names. Anything else will be noticed and considered the end of the parameter name. Anything following will be considered some modifier to be left out. There's a variety of possibilities that this might be invalid for. Are any other characters allowed in swagger variable names? Can similar modifiers be before the parameter name?

I'm having trouble finding any documentation on the templating that swagger is supposed to support. For example, under what circumstances is this colon:regex syntax is valid? I don't wanna dive too deep without having a better understanding of the templating involved. Any input/direction here would be great.

macjohnny commented:
@defmonk0 one more character that should be valid is the dollar sign `$`. 
Currently, however, it is replaced with an underscore `_` 
here
https://github.com/swagger-api/swagger-codegen/blob/master/modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/AbstractTypeScriptClientCodegen.java#L151
using this method here
https://github.com/swagger-api/swagger-codegen/blob/master/modules/swagger-codegen/src/main/java/io/swagger/codegen/DefaultCodegen.java#L3495

This leads e.g. to https://github.com/swagger-api/swagger-codegen/issues/7484

Maybe we could explicitly look for the colon `:` to find the end of the variable name, allowing any other character.

@wing328 do you have any docs to point us to in order to find out more about the valid syntax in path parameters?

macjohnny commented:
@defmonk0 I updated the code as I suggested above in https://github.com/swagger-api/swagger-codegen/pull/7479#issuecomment-360699167 to explicitly look for a colon `:` as a path parameter variable name terminator, if existing. 
Moreover, I fixed handling parameters that come after a parameter with a colon / special character by resetting 
`skipThisAndSubsequentChars = false;` in https://github.com/swagger-api/swagger-codegen/pull/7479/files#diff-ad60e100e050d5ef52943803a2adc5fdR277

With this changes, I could successfully test with the following swagger definition:
```
  /pet/{birthday}/foo/{id:.+}/{bla_bla}/bar:
    get:
      tags:
        - pet
      operationId: findPetsByStatusfoo
      produces:
        - application/xml
        - application/json
      parameters:
        - name: birthday
          in: path
          required: true
          type: string
          format: date-time
        - name: id
          in: path
          required: true
          type: string
        - name: bla_bla
          in: path
          required: true
          type: string
          format: date-time
      responses:
        '200':
          schema:
            type: array
            items:
              $ref: '#/definitions/Pet'
```

macjohnny commented:
@arjuin could you please test the newest code again and confirm this resolves issue #7476 ?
@inthegarage could you please test the newest code again and confirm this resolves issue #7302 ?

defmonk0 commented:
@macjohnny I definitely think that's a bit safer/future-proof, and the example looks good. If William doesn't have any particular input or documentation for additional possibilities that this misses, I think it should be fine. 👍

macjohnny commented:
@defmonk0 thanks for your response. would you mind reviewing the code with the github review feature in the top right corner?
![review](https://user-images.githubusercontent.com/5589029/35446484-2bd57902-02b5-11e8-925c-d96c78454cf4.png)

TiFu commented on modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/TypeScriptAngularClientCodegen.java:
```diff
@@ -246,6 +246,8 @@ public void postProcessParameter(CodegenParameter parameter) {
             StringBuilder pathBuffer = new StringBuilder();
             StringBuilder parameterName = new StringBuilder();
+            /** Whether to skip characters inside curly braces */
+            Boolean skipThisAndSubsequentChars = false;
```
Maybe use the primitive `boolean` type here and change the comment to the actual function of this variable?
Below it's not only used to skip chars in curly braces, but also to skip chars after a colon.

macjohnny replied:
@TiFu done, I updated the commend and changed the type to `boolean`

TiFu commented on samples/client/petstore/typescript-angular-v2/npm/tsconfig.json:
```diff
@@ -5,7 +5,7 @@
         "suppressImplicitAnyIndexErrors": true,
         "target": "es5",
-        "module": "es6",
+        "module": "commonjs",
```
Is this change correct? Seems like it was introduced in another PR.

macjohnny replied:
@TiFu yes, that's right, I guess this pops up since I had to re-generate all typescript files.

inthegarage commented:
@macjohnny Yes I agree this change is fine and code produced will work in the other scenario. Sadly I cannot test the change at the moment as other changes in 2.4.0-SNAPSHOT have broken the generation of the code. I will document these issues in due course.

macjohnny commented:
@inthegarage thanks for your feedback. could you specify what other issues you have with the current snapshot version? maybe we could include a quick fix in this PR

inthegarage commented:
@macjohnny 
I will add further issues here:
https://github.com/swagger-api/swagger-codegen/issues/7535

macjohnny commented:
@wing328 the issue reported by @inthegarage does not seem related to this PR.
Since this PR has been reviewed, it should be ready to be merged.

inthegarage commented:
@macjohnny  no it's not related. I should get time to review the other issue later.

wing328 commented:
If no further comment/question on this PR, I'll merge it on coming Sunday.

cc @TiFu @taxpon @sebastianhaas @kenisteward @Vrolijkx

macjohnny commented:
Following the discussion in https://github.com/swagger-api/swagger-codegen/issues/7302, this PR should not be merged yet.

macjohnny commented:
@wing328 the PR was reverted to comply with the discussion in #7302, so it should be ready to be merged now

macjohnny commented:
@gualtierim would you like to port this change to the typescript-inversify generator, since you use the same logic there?

gualtierim commented:
@macjohnny yes I will port this change

macjohnny commented:
@wing328 can you merge this?

macjohnny commented:
@wing328 the conflicts have been resolved, could you please merge this?

wing328 commented:
@macjohnny I've merged the PR into master. Thanks for the contribution.
